### PR TITLE
Fixes #22052 - Added help text in Templates tab

### PR DESCRIPTION
**Pull request by apuntamb**

### CODE REVIEW

**Comment by xprazak2** [comment]
Looking good visually:


The test failures are probably related.

File: app/views/operatingsystems/_templates.html.erb [hunk 1]
```diff
@@ -1,2 +1,2 @@
-<%= selectable_f f,:provisioning_template_id, @operatingsystem.provisioning_templates.where(:template_kind_id => f.object.template_kind_id).collect { |c| [c.name, c.id]}, { :include_blank => true }, { :label => _(f.object.template_kind.to_s) } %>
+<%= selectable_f f,:provisioning_template_id, @operatingsystem.provisioning_templates.where(:template_kind_id => f.object.template_kind_id).collect { |c| [c.name, c.id]}, { :include_blank => true }, { :label => _(f.object.template_kind.to_s), :help_inline => _(TemplatesHelper.template_help_text[f.object.template_kind.try(:name)])}  %>
```
Comment (ares) [inline app/views/operatingsystems/_templates.html.erb]: note we usually use `:label_help` that puts small info icon and help is displayed after click, would it work here so we are consistent across the whole app?

**Comment by ares** [comment]
new version pushed fixed tests, failures are now unrelated

File: app/views/operatingsystems/_templates.html.erb (hunk 1 above)
Comment (apuntamb) [reply app/views/operatingsystems/_templates.html.erb]: Thanks for the guidelines here. New version pushed includes :label_help

File: app/helpers/templates_helper.rb [hunk 2]
```diff
@@ -46,4 +46,19 @@ def locked_warning(template)
...
   end
+
+  def self.template_help_text
+    {
+      "PXELinux" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using pxelinux.0 file"),
+      "PXEGrub" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using grub/bootx64.efi file"),
+      "PXEGrub2" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using grub2/grubx64.efi file"),
+      "iPXE" => N_("Used in {g,i}PXE environments in place of PXELinux"),
+      "provision" => N_("The main unattended installation file; e.g. Kickstart or Preseed"),
+      "finish" => N_("A post-install script used to take custom actions after the main provisioning is complete"),
```
Comment (ares) [inline app/helpers/templates_helper.rb]: we should probably mention that for kickstart system, this is only used for image based provisioning when the image does is not configured to use user-data while for preseed systems, this runs also during network based provisioning

File: app/helpers/templates_helper.rb [hunk 3]
```diff
@@ -46,4 +46,19 @@ def locked_warning(template)
...
+      "script" => N_("An arbitrary script, not used by default, useful for certain custom tasks"),
+      "user_data" => N_("Similar to a Finish script, this can be assigned to hosts built on user_data-capable images (e.g. Openstack, EC2, etc)"),
```
Comment (ares) [inline app/helpers/templates_helper.rb]: Openstack, EC2 are not images but providers, perhaps mention cloud-init as the tool that fetches user_data?

**Review by ares — DISMISSED** [review]
Thanks, it looks better now. I read the help text and have two minor suggestions. Also could you please take a look at foreman_discovery which also adds some template type? It probably needs to extend the helper method. See attached screenshot for what I had in mind. The discovery part of course does not block this PR.

**Comment by ares** [comment]
templates

**Comment by apuntamb** [comment]
@ares  Thanks for the detailed guidelines, started working on fixing the same. While trying to setup the discovery plugin, I could not view the **Discovery Kexec Template** in the UI. Is it the **smart_proxy_discovery** plugin that needs to be enabled for the same? If yes, how does the foreman_discovery understands that smart_proxy_discovery is enabled? If no, could you just let me know of how to find the Kexec template?

**Comment by apuntamb** [comment]
@ares please scratch my most recent comment, got that!

File: app/helpers/templates_helper.rb [hunk 4]
```diff
@@ -46,4 +46,19 @@ def locked_warning(template)
 
     alert(:class => 'alert-warning', :text => warning_text.html_safe)
   end
+
+  def self.template_help_text
+    {
+      "PXELinux" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using pxelinux.0 file"),
+      "PXEGrub" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using grub/bootx64.efi file"),
+      "PXEGrub2" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using grub2/grubx64.efi file"),
+      "iPXE" => N_("Used in {g,i}PXE environments in place of PXELinux"),
```
Comment (lzap) [inline app/helpers/templates_helper.rb]: gPXE is super old, I'd go with iPXE here.

File: app/helpers/templates_helper.rb (hunk 3 above)
Comment (lzap) [reply app/helpers/templates_helper.rb]: Also the checkbox must be checked in order to use user-data. This can get pretty long, maybe we can detect it?

File: app/helpers/templates_helper.rb (hunk 4 above)
Comment (apuntamb) [reply app/helpers/templates_helper.rb]: Sure, I'll make that change as well. Thanks!

File: app/helpers/templates_helper.rb (hunk 3 above)
Comment (ares) [reply app/helpers/templates_helper.rb]: I wouldn't worry about the length, it's hidden in tooltip now. I think it's better to just describe and don't change help dynamically.

File: app/helpers/templates_helper.rb (hunk 2 above)
Comment (lzap) [reply app/helpers/templates_helper.rb]: Perhaps:

A post-install script for cloud instances without "user data" flag set or preseed PXE systems. Connection is made via SSH, credentials or key must exist and inventory IP address must match.

File: app/helpers/templates_helper.rb (hunk 3 above)
Comment (lzap) [reply app/helpers/templates_helper.rb]: Perhaps:

Template with seed data for virtual or cloud instances with "user data" flag set, typically cloud-init or ignition format.

**Comment by lzap** [comment]
After chat with @apuntamb we agreed that for plugins it would be much nicer to have the helper text in the template_kinds database schema. This will allow plugins putting any text nicely, it can be coded in a way that it's all i18n friendly and texts can be reused across the app.

**Comment by apuntamb** [comment]
@ares your views on the above comment regarding helper text in the template_kinds database schema ?

**Comment by ares** [comment]
I think that's even better solution, just make sure to don't display the label_help if the kind does not have any help set and that the strings are correctly extracted for translations (probably somewhere at seeds).

File: lib/tasks/gettext.rake [hunk 5]
```diff
@@ -14,7 +14,7 @@ begin
 
     # redefine file globs for Foreman
     def files_to_translate
-      Dir.glob("{app,lib,config,locale,webpack}/**/*.{rb,erb,haml,slim,rhtml,js,rabl}")
+      Dir.glob("{app,db,lib,config,locale,webpack}/**/*.{rb,erb,haml,slim,rhtml,js,rabl}")
```
Comment (lzap) [inline lib/tasks/gettext.rake]: Please add just `db/seeds.d`.

File: lib/tasks/save_template_help_text.rake [hunk 6]
```diff
@@ -0,0 +1,10 @@
+desc 'Storing templates help text to database'
+task :save_templates_help => :environment do
+  puts 'Saving Templates Help Text'
+  TemplateKind.reset_column_information
+  TemplateKind.all.each do |template_kind|
+    template_kind.template_help_text = template_kind.default_template_help[template_kind.name]
+    template_kind.save
+  end
+end
+
```
Comment (lzap) [inline lib/tasks/save_template_help_text.rake]: I am not sure what this is for. Can you explain?

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+class AddHelpTextToTemplateKinds < ActiveRecord::Migration[5.1]
+  def up
+    add_column :template_kinds, :template_help_text, :text
+
+    Rake::Task['save_templates_help'].invoke
```
Comment (lzap) [inline db/migrate/20180820072858_add_help_text_to_template_kinds.rb]: This does not appear to be a good idea. I don't understand what this rake task does.

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class AddHelpTextToTemplateKinds < ActiveRecord::Migration[5.1]
+  def up
+    add_column :template_kinds, :template_help_text, :text
```
Comment (lzap) [inline db/migrate/20180820072858_add_help_text_to_template_kinds.rb]: I vote for "description" column name rather than "text".

File: app/models/template_kind.rb [hunk 9]
```diff
@@ -24,6 +24,21 @@ def self.default_template_labels
...
+      "PXEGrub2" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using grub2/grubx64.efi file"),
+      "iPXE" => N_("Used in iPXE environments in place of PXELinux"),
+      "provision" => N_("In case of Kickstart systems, this is only used for image based provisioning when the image is not configured to use user-data, while for Preseed systems, this also runs during network based provisioning"),
+      "finish" => N_("A post-install script for cloud instances without user data flag set or preseed PXE systems. Connection is made via SSH, credentials or key must exist and inventory IP address must match"),
+      "script" => N_("An arbitrary script, not used by default, useful for certain custom tasks"),
+      "user_data" => N_("Template with seed data for virtual or cloud instances with user data flag set, typically cloud-init or ignition format."),
+      "ZTP" => N_("Provisioning Junos devices (Junos 12.2+)"),
+      "POAP" => N_("Provisioning for switches running NX-OS using DHCP and TFTP via Foreman")
+    }
+  end
```
Comment (lzap) [inline app/models/template_kind.rb:40]: This would become:

```ruby
def description
  _(read_attribute(:description))
end
```

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+    Rake::Task['save_templates_help'].invoke
+  end
+
+  def down
+    remove_column :template_kinds, :template_help_text
+  end
```
Comment (lzap) [inline db/migrate/20180820072858_add_help_text_to_template_kinds.rb:4]: Use new Rails migration syntax `change`.

**Review by lzap — CHANGES_REQUESTED** [review]
I apologize if I was not clear. My idea was to modify the seed script `db/seeds.d/020-provisioning_templates_list.rb` and creating all kinds explicitly there with kind name and newly added description:

```ruby
TemplateKind.create("PXELinux", N_("And description")
# etc
```

You need to do this idempotent, this is just a pseudo code.

Then change the gettext rake task to also extract those strings and make sure the newly added description is translated.

Use this setting in `settings.yaml` to make sure you see translated messages in UI:

```
# Mark translated strings with X characters (for developers)
:mark_translated: true
```

File: lib/tasks/save_template_help_text.rake (hunk 6 above)
Comment (apuntamb) [reply lib/tasks/save_template_help_text.rake]: This rake task was to save the help text into the db and I'd invoked this through the migration.

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb (hunk 8 above)
Comment (apuntamb) [reply db/migrate/20180820072858_add_help_text_to_template_kinds.rb]: sure.

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb (hunk 10 above)
Comment (apuntamb) [reply db/migrate/20180820072858_add_help_text_to_template_kinds.rb:4]: sure.

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb (hunk 7 above)
Comment (apuntamb) [reply db/migrate/20180820072858_add_help_text_to_template_kinds.rb]: This was the migration that invoked the rake task (used for saving help text into template_kinds table).

File: lib/tasks/gettext.rake (hunk 5 above)
Comment (apuntamb) [reply lib/tasks/gettext.rake]: sure.

File: lib/tasks/save_template_help_text.rake (hunk 6 above)
Comment (lzap) [reply lib/tasks/save_template_help_text.rake]: Hmmm I think it's cleaner to simply put the texts into DB during seed and remove them from the codebase completly.

File: db/migrate/20180820072858_add_help_text_to_template_kinds.rb (hunk 7 above)
Comment (lzap) [reply db/migrate/20180820072858_add_help_text_to_template_kinds.rb]: Yeah remove this please.

File: app/models/template_kind.rb [hunk 11]
```diff
@@ -24,6 +24,10 @@ def self.default_template_labels
     }
   end
 
+  def description
+    _(read_attribute(:description))
```
[the remark this answers is not in the document]
Comment (lzap) [reply app/models/template_kind.rb]: @ares what is the trick to define read attribute in a way that it is transparently translated?

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 12]
```diff
@@ -93,14 +109,19 @@ def seeded_templates
       ]
     end
 
-    def kinds
-      kinds = {}
+    @@kinds ||= begin
```
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: Please keep the `kinds` method to prevent that huge change of the hash above. What I was suggesting was actually Ruby memoization design pattern: https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/ - simply just use `@kinds` as in the examples and keep using the method please.

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 13]
```diff
@@ -1,62 +1,78 @@
...
+      "PXEGrub2" => N_("Deployed to the TFTP server to ensure the Host boots the correct installer using grub2/grubx64.efi file"),
+      "iPXE" => N_("Used in iPXE environments in place of PXELinux"),
+      "provision" => N_("In case of Kickstart systems, this is only used for image based provisioning when the image is not configured to use user-data, while for Preseed systems, this also runs during network based provisioning"),
+      "finish" => N_("A post-install script for cloud instances without user data flag set or preseed PXE systems. Connection is made via SSH, credentials or key must exist and inventory IP address must match"),
+      "script" => N_("An arbitrary script, not used by default, useful for certain custom tasks"),
+      "user_data" => N_("Template with seed data for virtual or cloud instances with user data flag set, typically cloud-init or ignition format."),
+      "ZTP" => N_("Provisioning Junos devices (Junos 12.2+)"),
+      "POAP" => N_("Provisioning for switches running NX-OS using DHCP and TFTP via Foreman")
+    }
+    end
```
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: This could be memoized too, not a big deal tho.

**Review by lzap — CHANGES_REQUESTED** [review]
It looks good, implementation details.

File: app/models/template_kind.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (ares) [reply app/models/template_kind.rb]: I guess you could simply call `_(super)` here, but the key is that all description values would need to be extracted for translation using `N_()` which I think you're doing at def `self.template_description`. 

I just don't think translating on this level is a good idea. Shouldn't it be in view layer? Note that translation is determined by current user locale, when I call this method e.g. in migration where there's no user set, it does not make sense.

File: db/seeds.d/020-provisioning_templates_list.rb (hunk 12 above)
Comment (apuntamb) [reply db/seeds.d/020-provisioning_templates_list.rb]: Sure, got it.

File: app/models/template_kind.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (apuntamb) [reply app/models/template_kind.rb]: Okay sure, I can store the simple text without translation in db and later, translate it at view layer.

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 14]
```diff
@@ -1,5 +1,21 @@
 class ProvisioningTemplatesList
   class << self
+    def template_description
+      @template_description ||=
+      {
+        "PXELinux" => "Deployed to the TFTP server to ensure the Host boots the correct installer using pxelinux.0 file",
```
Comment (apuntamb) [inline db/seeds.d/020-provisioning_templates_list.rb]: Currently I'm not translating the help text at db level, but only at the view level. So, after running `rake gettext:find` task, I'm unable to see the translations in the foreman.po file. Am I missing something?
@ares @lzap your comments?

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 15]
```diff
@@ -94,13 +109,17 @@ def seeded_templates
...
+          return @kinds[type] if @kinds.key?(type)
+          @kinds[type] ||= TemplateKind.unscoped.find_or_create_by(name: type)
+          @kinds[type].description = template_description[@kinds[type].name]
+          @kinds[type].save
+          raise "Unable to create template kind: #{format_errors @kinds[type]}" if @kinds[type].errors.any?
+        end
+        @kinds
       end
-      kinds
     end
```
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: This is what I meant. But do not access `@kind` inside of the block, just use local variable and return it.

**Comment by ares** [comment]
@lzap could you take another look? seems @apuntamb updated the PR

[test foreman] [test katello] [test upgrade] [test prprocessor]

File: app/views/operatingsystems/_templates.html.erb [hunk 16]
```diff
@@ -1,2 +1,3 @@
-<%= selectable_f f,:provisioning_template_id, @operatingsystem.provisioning_templates.where(:template_kind_id => f.object.template_kind_id).collect { |c| [c.name, c.id]}, { :include_blank => true }, { :label => _(f.object.template_kind.to_s) } %>
+<%= selectable_f f,:provisioning_template_id, @operatingsystem.provisioning_templates.where(:template_kind_id => f.object.template_kind_id).collect { |c| [c.name, c.id]}, { :include_blank => true }, { :label => _(f.object.template_kind.to_s), :label_help => _(f.object.template_kind.description) }  %>
+
```
Comment (lzap) [inline app/views/operatingsystems/_templates.html.erb]: Nitpick: Two spaces at the end of the line 1, is the extra line needed? I am fine just checking.

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 17]
```diff
@@ -94,13 +109,17 @@ def seeded_templates
...
     def kinds
-      kinds = {}
-      TemplateKind.default_template_labels.keys.map(&:to_sym).each do |type|
-        kinds[type] = TemplateKind.unscoped.find_by_name(type)
-        kinds[type] ||= TemplateKind.unscoped.create(:name => type)
-        raise "Unable to create template kind: #{format_errors kinds[type]}" if kinds[type].nil? || kinds[type].errors.any?
+      @kinds ||= begin
+        kinds = {}
+        TemplateKind.default_template_labels.keys.map(&:to_sym).each do |type|
+          return kinds[type] if kinds.key?(type)
```
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: Can you explain this return? I believe that it should never trigger as you iterate through unique keys. Let's remove it?

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 18]
```diff
@@ -94,13 +109,17 @@ def seeded_templates
...
+          kinds[type] ||= TemplateKind.unscoped.find_or_create_by(name: type)
+          kinds[type].description = template_description[kinds[type].name]
+          kinds[type].save
+          raise "Unable to create template kind: #{format_errors kinds[type]}" if kinds[type].errors.any?
```
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: I'd prefer introducing a local variable `kind` and on the last line storing it `kinds[type] = kind` this is little bit weird. But it works if you prefer this.

Comment (lzap) [reply db/seeds.d/020-provisioning_templates_list.rb]: Can you actually use `save!` instead? It will trigger the exception for you. Let's improve this.

**Review by lzap — CHANGES_REQUESTED** [review]
It looks great! I have some minor comments and then I would like to propose better helper texts:

```ruby
      @template_description ||= {
        "PXELinux" => N_("Used when PXELinux loader is set, loads pxelinux.0 which loads content generated by this template."),
        "PXEGrub" => N_("Used when Grub2 loader is set, loads grub/bootx64.efi which loads content generated by this template."),
        "PXEGrub2" => N_("Used when PXELinux loader is set, loads grub/grubx64.efi which loads content generated by this template."),
        "iPXE" => N_("Used in iPXE environments."),
        "provision" => N_("Template for OS installer, for example kickstart, preseed or jumpstart. Depends on the operating system."),
        "finish" => N_("Post-install script for preseed-based or cloud instance. Connection is made via SSH, credentials or key must exist and inventory IP address must match. Only used when 'user data' is not set."),
        "script" => N_("An arbitrary script, must be manually downloaded using wget/curl."),
        "user_data" => N_("Template with seed data for virtual or cloud instances when 'user data' flag is set, typically cloud-init or ignition format."),
        "ZTP" => N_("Provisioning Junos devices (Junos 12.2+)."),
        "POAP" => N_("Provisioning for switches running NX-OS.")
      }
```

**Comment by lzap** [comment]
Also the issue number is incorrect, it must be: https://projects.theforeman.org/issues/24684

File: db/seeds.d/020-provisioning_templates_list.rb (hunk 17 above)
Comment (apuntamb) [reply db/seeds.d/020-provisioning_templates_list.rb]: This return was mentioned with the thought of running db:seed post the data is already in table (i.e. not the first time), so that if key in that hash already exists, it would not execute any further steps. But, I agree, we can remove it.

File: db/seeds.d/020-provisioning_templates_list.rb (hunk 18 above)
Comment (apuntamb) [reply db/seeds.d/020-provisioning_templates_list.rb]: Sounds great.

**Comment by apuntamb** [comment]
> Also the issue number is incorrect, it must be: https://projects.theforeman.org/issues/24684

@lzap The #24684 issue is the help text issue for kexec template in foreman_discovery. This issue number is for foreman default templates.

File: db/seeds.d/020-provisioning_templates_list.rb (hunk 17 above)
Comment (apuntamb) [reply db/seeds.d/020-provisioning_templates_list.rb]: Also, for this seed, I was planning to write tests in [seed_tests.rb](https://github.com/theforeman/foreman/blob/develop/test/unit/tasks/seeds_test.rb) .
Is that the correct? Any suggestions / ideas on test ?

Comment (lzap) [reply db/seeds.d/020-provisioning_templates_list.rb]: I don't think this feature requires a test - description missing is non-fatal.

**Comment by lzap** [comment]
Okay, but the referenced issue is named "Templates tab in Hosts>Operating Systems have confusing and misleading asterisks" as well as the BZ referenced from there. @ares or @iNecas is this expected bugfix? Shall we rename the issue?

**Review by lzap — DISMISSED** [review]
Ready to test

**Comment by lzap** [comment]
It works fine, but couple of questions or points:

* What are we gonna do about the `*` character? The original issue is about removing those, I still see them.
* These descriptions are really helpful, can you spread them a little bit? You can put them at:
  * Provisioning Templates index page (Kind column)
  * Provisioning Templates edit page (Type tab)

I think that's all :-)

**Comment by iNecas** [comment]
I agree with @lzap that the issue is not solved until the '*' is gone. The problem is the form takes it by default from activerecord validations (https://github.com/inecas/foreman/blob/d5288f2bf6d59fcd78745b02c0933e6e5da65460/app/models/os_default_template.rb#L6). I think passing the `required` option explicitly (looking at https://github.com/inecas/foreman/blob/d5288f2bf6d59fcd78745b02c0933e6e5da65460/app/helpers/form_helper.rb#L337) should help avoiding loading it from activerecord, which is wrong in this particular case.

**Comment by apuntamb** [comment]
Thank you @lzap and @iNecas for the comments. But, I would also like to throw some light on [#22052](https://projects.theforeman.org/issues/22052#note-1) mentioned by @ares . This suggests keeping all asterisks but add help to the form that explains the purpose.
So, what should be our next step? Removing all the asterisks or keeping them along with help text?

**Comment by apuntamb** [comment]
@lzap  Also, I had a query : on Provisioning Templates index page, should we remove the existing template kind name and insert template kind description in the Kind column ? If so, should we change `kinds[type] = kind` in `db/seeds.d/020-provisioning_templates_list.rb` to `kinds[type] = kind.description` ? or we do not need to make changes in the db for this ?

**Comment by lzap** [comment]
@apuntamb I think you are doing it too complicated, let's just remove the asterisks as they were confusing. Now we have also helper text which is great.

In regard to replacing columns with description, the description is much longer and this would probably break the table, I suggest to simply keep kind there and either add icon or simply hover text with description when you hover the mouse on the kind text.

There are conflicts you need to rebase.

File: db/seeds.d/020-provisioning_templates_list.rb [hunk 19]
```diff
@@ -98,13 +113,16 @@ def seeded_templates
...
+        TemplateKind.default_template_labels.keys.map(&:to_sym).each do |type|
+          kind ||= TemplateKind.unscoped.find_or_create_by(name: type)
+          kind.description = template_description[kind.name]
+          kind.save!
+          kinds[type] = kind
+        end
+        kinds
       end
-      kinds
     end
```
Comment (lzap) [inline db/seeds.d/020-provisioning_templates_list.rb]: This block has been moved to `db/seeds.d/020-provisioning_templates_list.rb` in d689adb5ebc2e93d1fa9944642d08a71a71f931b so you need to rebase on top of that and put your changes there. Easiest way is to make a copy of this file, then rebase and "use deleted file" and then change the new file manually.

**Comment by apuntamb** [comment]
@lzap The latest patch now adds help text on Hosts>Operating Systems>Templates using the label_help with the `*` no longer present. Also, the `Kind` column on Provisioning Templates index page, now shows the same help text on mouse hover. If we select a template from the same page and click on `Type` tab, the help text specific to only that Template will be shown.

Please let me know for anything that I've missed.

File: app/models/template_kind.rb [hunk 20]
```diff
@@ -24,6 +24,21 @@ def self.default_template_labels
     }
   end
 
+  def self.template_description
```
Comment (lzap) [inline app/models/template_kind.rb]: Nitpick: To reflect the `default_template_labels` method name above this could be `default_template_descriptions`.

File: app/views/provisioning_templates/index.html.erb [hunk 21]
```diff
@@ -27,7 +27,12 @@
                                          :permission => 'edit_provisioning_templates') %>
         </td>
         <td class="ellipsis"><%= combination provisioning_template %></td>
-        <td><%= provisioning_template.try(:template_kind) %></td>
+        <td>
+          <% tmpl = provisioning_template.try(:template_kind) %>
+            <% if tmpl %>
+              <%= link_to tmpl, provisioning_templates_path(:search => "kind = #{tmpl.name}"), :title => "#{TemplateKind.template_description[tmpl.name]}" %>
+            <% end %>
+        </td>
```
Comment (lzap) [inline app/views/provisioning_templates/index.html.erb:34]: The indent of line 32 is not needed actually. You can also do `if tmpl = xxx` if Rubocop won't complain.

File: app/models/os_default_template.rb [hunk 22]
```diff
@@ -3,7 +3,6 @@ class OsDefaultTemplate < ApplicationRecord
   belongs_to :template_kind
   belongs_to :operatingsystem
 
-  validates :provisioning_template_id, :presence => true
```
Comment (lzap) [inline app/models/os_default_template.rb]: Is this change intentional? Can you explain?

**Review by lzap — CHANGES_REQUESTED** [review]
Couple of comments, there is a validator removed, not sure why.

File: app/models/os_default_template.rb (hunk 22 above)
Comment (apuntamb) [reply app/models/os_default_template.rb]: I observed that the templates are mandatory in Hosts>OS>Templates because of this validation and hence, asterisks are shown. In order to make the template fields non-mandatory (i.e. removing asterisks), this change was done.

Comment (ares) [reply app/models/os_default_template.rb]: I think we shouldn't remove validations, you can override the asterisk display by passing `:required => false` to the form helper

File: app/views/provisioning_templates/index.html.erb (hunk 21 above)
Comment (ares) [reply app/views/provisioning_templates/index.html.erb:34]: this is still valid concern, if you will combine if and assignment, make sure you wrap it in parentheses, like this
`<% if (tmpl == provisioning_template.try(:template_kind)) %>`

File: app/models/os_default_template.rb (hunk 22 above)
Comment (apuntamb) [reply app/models/os_default_template.rb]: @ares thanks, will make the changes.

File: app/views/provisioning_templates/_custom_tabs.html.erb [hunk 23]
```diff
@@ -3,7 +3,7 @@
 
   <div id="kind_selector" <%= display? @template.snippet %>>
     <%= select_f f, :template_kind_id, TemplateKind.all, :id, :to_s, {:include_blank => true, :selected => @template_kind_id},
-                 {:disabled => @template.locked?, :label => _("Type")} %>
+                 {:disabled => @template.locked?, :label => _("Type"), :label_help => _(f.object.template_kind.description)} %>
```
Comment (lzap) [inline app/views/provisioning_templates/_custom_tabs.html.erb]: Can you actually remove this one? It does not make sense for this kind of edit form, sorry my bad. Other than that it is good to merge.

**Review by lzap — CHANGES_REQUESTED** [review]
One last thing.

File: test/unit/tasks/seeds_test.rb [hunk 24]
```diff
@@ -205,4 +205,12 @@ def seed(*seed_files)
...
     assert_equal [], view_permissions - Role.unscoped.find_by_name('Viewer').permissions
   end
+
+  test "adds description to template kind" do
+    seed('070-provisioning_templates.rb')
+    tmpl_kind = TemplateKind.unscoped.find_by_name('provision')
+    tmpl_kind.description = 'test description provision kind'
+    tmpl_kind.save!
+    assert TemplateKind.unscoped.find_by_name('provision')
+  end
```
Comment (lzap) [inline test/unit/tasks/seeds_test.rb:213]: This test does not seem to be right, you want to make sure that the seed script 070 actually creates description. So simplify it a bit, just:

```ruby
seed('070-provisioning_templates.rb')
tmpl_kind = TemplateKind.unscoped.find_by_name('provision')
assert_equal "Expected description", tmpl_kind.description
```

**Review by lzap — CHANGES_REQUESTED** [review]
One comment on the new test :-)

File: test/unit/tasks/seeds_test.rb (hunk 24 above)
Comment (apuntamb) [reply test/unit/tasks/seeds_test.rb:213]: Thanks for the inputs. I have modified the test but, foreman tests seem to go red because of [https://ci.theforeman.org/job/test_develop_pr_core/19834/](https://ci.theforeman.org/job/test_develop_pr_core/19834/) . I had rebased and sent but still, this seems to be an issue. Is this being caused by my PR ?

File: test/models/template_kind_test.rb [hunk 25]
```diff
@@ -17,4 +17,11 @@ class TemplateKindTest < ActiveSupport::TestCase
     kind = FactoryBot.build_stubbed(:template_kind)
     assert_equal kind.name, kind.to_s
   end
+
+  test '#default_template_descriptions returns description of every template kind' do
+    tmpl = TemplateKind.default_template_descriptions
+    tmpl.keys.each do |type|
+      assert_not_empty tmpl[type]
+    end
+  end
```
Comment (lzap) [inline test/models/template_kind_test.rb:26]: This kinda test a fixture, but it's fine.

File: test/unit/tasks/seeds_test.rb (hunk 24 above)
Comment (lzap) [reply test/unit/tasks/seeds_test.rb:213]: Yes, these are non-related failures. This is a PASS!

**Review by lzap — APPROVED** [review]
Thanks Ready To Merge.
